### PR TITLE
mini lcm rich multi strings

### PR DESCRIPTION
**Pull request by hahn-kev**
actually trying to support runs in multi strings using an object with fields for data

## Summary by CodeRabbit

- **New Features**
  - Advanced rich text support has been introduced across various content fields—including definitions, notes, sentences, and translations—enabling enhanced formatting such as styling, alignment, and color customization.
  - The data input and display mechanisms have been streamlined, resulting in more consistent and visually rich text rendering and a more robust editing experience for users.

### CODE REVIEW

**Comment by hahn-kev** [comment]
this is really only defining the in memory model for rich multi strings and mapping them from LCM and back. Then we need to figure out migrating existing data, shouldn't be too bad but I'd like that to be a separate PR as this one is already getting large.

File: backend/FwLite/FwDataMiniLcmBridge/Api/UpdateProxy/UpdateEntryProxy.cs [hunk 1]
```diff
@@ -75,3 +76,28 @@ public override MultiString Copy()
...
+                     throw new ArgumentException($"unable to convert value {value?.GetType().Name ?? "null"} to string",
+                         nameof(value));
+        if (key is WritingSystemId keyWs)
+        {
+            proxy.Add(keyWs, valStr);
+        }
+        else if (key is string keyStr)
+        {
+            proxy.Add(keyStr, valStr);
+        }
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/FwLite/FwDataMiniLcmBridge/Api/UpdateProxy/UpdateEntryProxy.cs:97]: We need to cover both because the key variable has the type of object. If because of that you can't cast a WsId to a string or string to WsId. It's messy and an edge case, but that's why as the runtime type could be either and this is the only way to handle that.

File: backend/FwLite/MiniLcm/Models/RichMultiString.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    public static bool IsValidRichString(string value)
+    {
+        var doc = new HtmlDocument();
+        doc.LoadHtml(value);
+        if (doc.ParseErrors.Any()) return false;
+        if (doc.DocumentNode.ChildNodes.Count == 0) return false;
+        foreach (var childNode in doc.DocumentNode.ChildNodes)
+        {
+            if (childNode.NodeType != HtmlNodeType.Element) return false;
+            foreach (var childNodeChildNode in childNode.ChildNodes)
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/FwLite/MiniLcm/Models/RichMultiString.cs:21]: I guess? but that would be up to the parser, so js can do whatever it wants but it'll get parsed again in the C# side.

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+            FwTextPropType.ktptPadBottom => span => ref span.PadBottom,
+            FwTextPropType.ktptBorderTop => span => ref span.BorderTop,
+            FwTextPropType.ktptBorderBottom => span => ref span.BorderBottom,
+            FwTextPropType.ktptBorderLeading => span => ref span.BorderLeading,
+            FwTextPropType.ktptBorderTrailing => span => ref span.BorderTrailing,
+            FwTextPropType.ktptBulNumScheme => span => ref span.BulNumScheme,
+            FwTextPropType.ktptBulNumStartAt => span => ref span.BulNumStartAt,
+            FwTextPropType.ktptDirectionDepth => span => ref span.DirectionDepth,
+            FwTextPropType.ktptKeepWithNext => span => ref span.KeepWithNext,
+            FwTextPropType.ktptKeepTogether => span => ref span.KeepTogether,
```
Comment (myieye) [inline backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:94]: I see there are a handful of properties that are handled as ints both in Lcm and your code, but they're marked as FwTextPropVar.ktpvEnum and in some cases they're commented as being "ComBools". So, most of them are conceptually nullable booleans. I think the rule is just 0 == false, 1 == true everything else could probably be viewed as null/unspecified.

Right now we're primarily just trying to round trip everything, but that might be something worth improving now or later by mapping them to bools or enums.

The ones I see are: ktptKeepTogether, ktptKeepWithNext, ktptRightToLeft, ktptWidowOrphanControl, ktptBulNumScheme, ktptHyphenate, ktptSetRowDefaults.

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        FwTextPropType.ktptCharStyle,
+        FwTextPropType.ktptFontFamily,
+        FwTextPropType.ktptNamedStyle,
+        FwTextPropType.ktptBulNumTxtBef,
+        FwTextPropType.ktptBulNumTxtAft,
+        FwTextPropType.ktptFontVariations,
+        FwTextPropType.ktptParaStyle,
+        FwTextPropType.ktptTabList,
+        FwTextPropType.ktptCustomBullet,
+        FwTextPropType.ktptBulNumFontInfo,
```
Comment (myieye) [inline backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:60]: This looks like a pretty teriffying value under the hood 😱 😅

**Review by myieye — APPROVED** [review]
I didn't find any issues. So, hopefully that's a good thing 😅.
Nice work on this!

File: backend/FwLite/MiniLcm/Models/RichMultiString.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (myieye) [reply backend/FwLite/MiniLcm/Models/RichMultiString.cs:21]: Yeah, it doesn't really matter.

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs (hunk 4 above)
Comment (hahn-kev) [reply backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:60]: Yeah that one and WsStyle

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs (hunk 3 above)
Comment (hahn-kev) [reply backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:94]: Yeah that's a good point, I figure we'll do this until we want to actually use those props. The annoying thing is that a lot of these aren't even used anymore and I don't know which is which. All I know for sure is that NamedStyle, Ws, and ObjData are used heavily.

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+            FwTextPropType.ktptBulNumFontInfo => span => ref span.BulNumFontInfo,
+            FwTextPropType.ktptWsStyle => span => ref span.WsStyle,
+            FwTextPropType.ktptFieldName => span => ref span.FieldName,
+            _ => throw new ArgumentException($"property type {type} is not a string")
+        };
+    }
+
+    public static void WriteToSpan(RichSpan span, ITsTextProps textProps, Func<int?, WritingSystemId?> wsIdLookup)
+    {
+        for (int i = 0; i < textProps.IntPropCount; i++)
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:129]: done, I like that it throws now, but I don't like how the switch looks, but it's worth it I think

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+        {
+            null => null,
+            RichTextSpellingMode.Normal => (int)SpellingModes.ksmNormalCheck,
+            RichTextSpellingMode.DoNotCheck => (int)SpellingModes.ksmDoNotCheck,
+            RichTextSpellingMode.ForceCheck => (int)SpellingModes.ksmForceCheck,
+            _ => (int?)spellCheck
+        };
+    }
+
+    public static void WriteToTextProps(RichSpan span,
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:271]: so I tried this, but due to the duplicated/reused prop types this is much harder and the code ended up being really ugly and I still didn't handle all the edge cases. So I decided to revert it. At the end of the day we only support the props we've defined on RichSpan, I suppose we could write a simple test to ensure if a new prop type is added that we have a failing test.

File: backend/FwLite/FwDataMiniLcmBridge.Tests/RichTextTests.cs [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+        var allTypes = Enum.GetValues<FwTextPropType>();
+        var notTested = allTypes
+            .Except(excludedTypes)
+            .Except(testedTypes)
+            .ToArray();
+        if (notTested.Length > 0) output.WriteLine($"Not tested: {string.Join(", ", notTested)}");
+        notTested.Should().BeEmpty("values should be tested, {0} are not", notTested.Length);
+
+        //all these types are duplicated so must be in both string and int tested types
+        HashSet<FwTextPropType> duplicatedTypes =
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/FwLite/FwDataMiniLcmBridge.Tests/RichTextTests.cs:486]: I'm not sure if this is what you had in mind, but all we really need is the list of duplicated types so we can ensure that they're mapped in both strings and ints and they are.

File: backend/FwLite/FwLiteProjectSync.Tests/SyncTests.cs [hunk 8]
```diff
@@ -19,16 +19,16 @@ public class SyncTests : IClassFixture<SyncFixture>, IAsyncLifetime
     {
         Id = Guid.NewGuid(),
         LexemeForm = { Values = { { "en", "Apple" } } },
-        Note = { Values = { { "en", "this is a test note" } } },
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/FwLite/FwLiteProjectSync.Tests/SyncTests.cs:22]: Yeah I agree, not today though lol

**Comment by hahn-kev** [comment]
Alright I applied a bunch of your feedback, nice find that GetValues returns duplicates!

I also simplified `RichTextObjectData` as most use cases will benefit from it just being an object rather than abstract, it also simplifies json serialization.

File: backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs [hunk 9]
```diff
@@ -133,7 +134,15 @@ public static void WriteToSpan(RichSpan span, ITsTextProps textProps, Func<int?,
             if (IntProps.Contains(propType))
             {
                 MapToDelegateInt(propType).Invoke(span) = value;
+                //these types have values mapped from variation to unit
+                if (propType is FwTextPropType.ktptFontSize or FwTextPropType.ktptOffset or FwTextPropType.ktptLineHeight)
```
Comment (myieye) [inline backend/FwLite/FwDataMiniLcmBridge/Api/RichTextMapping.cs:138]: I would have put the `if (IntProps.Contains)` at the end of the switch. Then you wouldn't have to special case these.
